File: dsa/linked_list/876_Middle_of_the_Linked_List.java
package linked_list;
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
class Solution {
    public ListNode middleNode(ListNode head) {
        // 2 first condition to check if listnode is null or listnode just contain 1 node
        // 2 pointer, one fast, one slow
        if (head == null){
            return null;
        }

        ListNode slowPointer = head;
        ListNode fastPointer = head;

        while (fastPointer.next != null){
            fastPointer = fastPointer.next.next;
            slowPointer = slowPointer.next;
            if (fastPointer == null){
                break;
            }
        }
        return slowPointer;
    }
}

File: dsa/7.linked_list/876_Middle_of_the_Linked_List.py
import typing as t
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
class Solution:
    def middleNode(self, head: t.Optional[ListNode]) -> t.Optional[ListNode]:
        slow_pointer: ListNode = head
        fast_pointer: ListNode = head

        while fast_pointer and fast_pointer.next:
            slow_pointer = slow_pointer.next
            fast_pointer = fast_pointer.next.next
        
        return slow_pointer